### PR TITLE
Fixed mysql_user idempotency for long privilege lists.

### PR DESCRIPTION
**Pull request by apollo13**
##### SUMMARY
Fixes ansible/ansible#68044

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
mysql_user

### CODE REVIEW

**Comment by apollo13** [comment]
@Andersson007 where would I put the changelog fragment in the collections repo?

**Comment by gundalow** [comment]
@apollo13 `changelogs/fragments` (you'll need to create the directory as we don't have any yet)
Thank you for the PR :)

**Comment by Andersson007** [comment]
@apollo13 thanks for moving this to here!
could you please remind me whether all things were ok with CI tests in the ansible/ansible, weren't them?

**Comment by apollo13** [comment]
They were good aside from an apparently flaky test: https://app.shippable.com/github/ansible/ansible/runs/161886/69/tests

**Comment by Andersson007** [comment]
@apollo13 yep, it's no problem, happens sometimes, unrelated.
@bmalynovytch could you please look at this one, we had the same PR with CI passed in ansible/ansible

**Comment by bmalynovytch** [comment]
Would you mind adding the link of the previous PR ?
Thx 🙂

**Comment by apollo13** [comment]
@bmalynovytch Sorry, it was in the original ticket, but here for reference: https://github.com/ansible/ansible/issues/68044

File: plugins/modules/database/mysql/mysql_user.py [hunk 1]
```diff
@@ -561,14 +561,14 @@ def pick(x):
         res = re.match("""GRANT (.+) ON (.+) TO (['`"]).*\\3@(['`"]).*\\4( IDENTIFIED BY PASSWORD (['`"]).+\\6)? ?(.*)""", grant[0])
         if res is None:
             raise InvalidPrivsError('unable to parse the MySQL grant string: %s' % grant[0])
-        privileges = res.group(1).split(", ")
-        privileges = [pick(x) for x in privileges]
+        privileges = res.group(1).split(",")
```
Comment (bmalynovytch) [inline plugins/modules/database/mysql/mysql_user.py:564]: By default, this regex won't match multiline.

Extract from https://docs.python.org/3/library/re.html:
```
[...]
The special characters are:

'.'
(Dot.) In the default mode, this matches any character except a newline. If the DOTALL flag has been specified, this matches any character including a newline.
[...]
```

I therefore suppose that the `split` + `strip` + `join` may break if a double-keyword is split in two lines, as the priv would be re-aggregated as two separate keywords.

Comment (apollo13) [reply plugins/modules/database/mysql/mysql_user.py:564]: I don't think it should match multiple lines, mysql actually produces __multiple__ `GRANT` statements:

```
+-----------------------------------------------------------------------------------------------------------------------------------------------------+
| Grants for replication_admin@%                                                                                                                      |
+-----------------------------------------------------------------------------------------------------------------------------------------------------+
| GRANT SELECT, RELOAD, SHUTDOWN, PROCESS, FILE, SUPER, EXECUTE, REPLICATION SLAVE, REPLICATION CLIENT, CREATE USER ON *.* TO `replication_admin`@`%` |
| GRANT BACKUP_ADMIN,CLONE_ADMIN,PERSIST_RO_VARIABLES_ADMIN,SYSTEM_VARIABLES_ADMIN ON *.* TO `replication_admin`@`%` WITH GRANT OPTION                |
| GRANT INSERT, UPDATE, DELETE ON `mysql`.* TO `replication_admin`@`%` WITH GRANT OPTION                                                              |
| GRANT ALL PRIVILEGES ON `mysql_innodb_cluster_metadata`.* TO `replication_admin`@`%` WITH GRANT OPTION                                              |
| GRANT ALL PRIVILEGES ON `mysql_innodb_cluster_metadata_bkp`.* TO `replication_admin`@`%` WITH GRANT OPTION                                          |
| GRANT ALL PRIVILEGES ON `mysql_innodb_cluster_metadata_previous`.* TO `replication_admin`@`%` WITH GRANT OPTION                                     |
+-----------------------------------------------------------------------------------------------------------------------------------------------------+
```
As you can see there, the first two lines are for replication_admin

Comment (bmalynovytch) [reply plugins/modules/database/mysql/mysql_user.py:564]: My bad (covid got me 🙄🤒 ), when I read your PR, your long grant was on multiple lines but I didn't notice that there was no line numbers, meaning it was on a single line.
I was wondering why this old regex never crashed... I now know why 😄

Comment (Andersson007) [reply plugins/modules/database/mysql/mysql_user.py:564]: @bmalynovytch get well soon!

Comment (bmalynovytch) [reply plugins/modules/database/mysql/mysql_user.py:564]: Thanks ! Still 4/5 days before doing well again !

Comment (Andersson007) [reply plugins/modules/database/mysql/mysql_user.py:564]: ohh..

**Comment by felixfontein** [comment]
Thanks everyone! Feel free to backport this to ansible/ansible if you want this to appear in 2.9.x.

**Comment by Andersson007** [comment]
@apollo13 could you backport it?

File: plugins/modules/database/mysql/mysql_user.py (hunk 1 above)
Comment (apollo13) [reply plugins/modules/database/mysql/mysql_user.py:564]: Ouch, all the best and sorry for pinging you.

**Comment by apollo13** [comment]
@Andersson007 Jupp will give it a try later :)

**Comment by apollo13** [comment]
Added backport in https://github.com/ansible/ansible/pull/68620

**Comment by Andersson007** [comment]
@apollo13 i'll ask someone to trigger the tests agains the backport pr (one unrelated unstable there), everything will be ok
and there are still a lot of mysql-related issues/pr which can be reviewed/rewritten:)
i can help with review
thanks!

**Pull request merged**